### ISSUE TITLE
[QST] Translating pandas code to dask-cudf code

### ISSUE THREAD

**Issue by aclifton314** [rapidsai/cudf#8395]
**System Info**
System: Red Hat Enterprise Linux Server release 7.7 (Maipo)
CUDA: 10.2
dask-cuda: 0.19.0
python: 3.7.10
GPU: 4 NVIDIA Titan V-100

**Question**
I'm struggling with making a translation between a `pandas` implementation of longest common subsequence and a `dask-cudf` implementation. Below is the `pandas` implementation:
```python
import spacy
import pandas as pd

if __name__ == "__main__":
    nlp = spacy.load('en_core_web_sm')
    spacy.require_gpu()

    #Takes two string inputs and calculates the longest common subsequence between them, returning that longest common
    #subsequence.
    #Note, it's not mandatory that spacy be used (i.e. doca = nlp(seq_a)) to run the for loops.
    def lcsubseq(seq_a, seq_b):
        doca = nlp(seq_a)
        docb = nlp(seq_b)

        # generate matrix of length of longest common subsequence for substrings of both words
        lengths = [[0] * (len(docb) + 1) for _ in range(len(doca) + 1)]
        for ii, tok_a in enumerate(doca):
            for jj, tok_b in enumerate(docb):
                if tok_a.text.lower() == tok_b.text.lower():
                    lengths[ii + 1][jj + 1] = lengths[ii][jj] + 1
                else:
                    lengths[ii + 1][jj + 1] = max(lengths[ii + 1][jj], lengths[ii][jj + 1])

        # read a substring from the matrix
        result = []
        docb_len = len(docb)
        for i in range(1, len(doca) + 1):
            if lengths[i][docb_len] != lengths[i - 1][docb_len]:
                result.append(doca[i - 1].text)
        return ' '.join(result)


    #since seq_df contains the preprocessed sequences as its indices and column names, this seq_df is upper diagonal.
    #We set the diagonal elements to 1 manually. I haven't introduced logic to skip over the lower diagonal so there
    #is some redundancy. However, the calculation for seq_a != seq_b and seq_a_index != seq_b_index is given in the else
    #portion of the loop.
    def make_lcsubseq_df(seq_df):
        cols = seq_df.columns.to_list()
        idx_counter = 0
        for seq_a in seq_df.index.to_list():
            for col_idx in range(idx_counter, len(cols)):
                if col_idx == idx_counter:
                    seq_b = cols[col_idx]
                    seq_df[seq_b][seq_a] = 1
                else:
                    seq_b = cols[col_idx]
                    seq_df[seq_b][seq_a] = lcsubseq(seq_a, seq_b)
        return seq_df


    #convert a pandas text series to a spacy document and store the tokens that are not punctuation or spaces.
    def process_text(text_series, process_method: str, batch_size=256):
        if 'no_punct_lower' in process_method:
            docs = nlp.pipe(text_series, batch_size=batch_size, disable=['parser', 'ner'])
            new_strs_list = []
            for doc in docs:
                proc_list = []
                for token in doc:
                    if token.pos_ != 'PUNCT' and token.pos_ != 'SPACE':
                        proc_list.append(token.text.lower())
                new_strs_list.append(' '.join(proc_list))
            return pd.Series(new_strs_list)
        else:
            return text_series

    #read in text data
    data_path = '/path/to/data.csv'
    data_df = pd.read_csv(data_path, encoding='ISO-8859-1')

    #set a run_type and grab the corresponding column from the data_df.
    run_type = 'lemma_no_punct_lower'
    data_series = data_df['lemmatized']

    #pre-process the text
    new_strs = process_text(data_series, run_type)

    #create and empty dataframe whose indices and columns are the preprocessed text.
    df = pd.DataFrame(columns=new_strs, index=new_strs)

    #calculate the longest common subsequence for each index,column pair.
    lcs_df = make_lcsubseq_df(df)
```

As it stands, the `pandas` implementation takes a long time on CPU (greater than 4 days). Since I have 4 GPUs available to me, I wanted to utilize them with `dask-cudf` to make the calcualtion. What I'm struggling with is how to properly do this such that the calculation is sped up and distributed across the GPUs. Here is what I have thus far:
```python
import dask_cudf
import cudf
import spacy
import subprocess
from dask_cuda import LocalCUDACluster
from dask.distributed import Client
import pandas as pd

if __name__ == "__main__":
    print('loading spacy model...')
    nlp = spacy.load('en_core_web_sm')
    spacy.require_gpu()

    # convert a pandas text series to a spacy document and store the tokens that are not punctuation or spaces.
    def process_text(text_series, process_method: str, batch_size=256):
        if 'no_punct_lower' in process_method:
            docs = nlp.pipe(text_series, batch_size=batch_size, disable=['parser', 'ner'])
            new_strs_list = []
            for doc in docs:
                proc_list = []
                for token in doc:
                    if token.pos_ != 'PUNCT' and token.pos_ != 'SPACE':
                        proc_list.append(token.text.lower())
                new_strs_list.append(' '.join(proc_list))
            return pd.Series(new_strs_list)
        else:
            return text_series

    def setup_dask():
        cmd = "hostname --all-ip-addresses"
        process = subprocess.Popen(cmd.split(), stdout=subprocess.PIPE)
        output, error = process.communicate()
        IPADDR = str(output.decode()).split()[0]

        cluster = LocalCUDACluster(ip=IPADDR)
        client = Client(cluster)
        return cluster, client

    #setup gpu cluster and client
    gpu_cluster, dask_client = setup_dask()

    # read in text data
    data_path = '/path/to/data.csv'
    data_df = pd.read_csv(data_path, encoding='ISO-8859-1')

    # set a run_type and grab the corresponding column from the data_df.
    run_type = 'lemma_no_punct_lower'
    data_series = data_df['lemmatized']

    # pre-process the text
    new_strs = process_text(data_series, run_type)

    # create and empty dataframe whose indices and columns are the preprocessed text.
    df = pd.DataFrame(columns=new_strs, index=new_strs)

    # write the pandas df to a csv file then read it in as a dask_cudf dataframe.
    df.to_csv('./tmp_csv.csv')
    data_dask_df = dask_cudf.read_csv('./tmp_csv.csv', encoding='ISO-8859-1')

```
You can see that `make_lcsubseq_df()` and `lcsubseq()` are not implemented, mostly because I'm not sure how to proceed from this point in a similar way as the `pandas` implementation. Any advice on how to make the translation from `pandas` to `dask-cudf` is very much appreciated!

**Comment by quasiben** [comment]
If possible, can you share a bit more of the context around this problem ? Are you working on a benchmark, part of a larger research project ?

**Comment by aclifton314** [comment]
@quasiben Sure! This is part of a larger project. I'm doing some NLP research trying to create a text classifier. However, as part of the data exploration effort, I wanted to see what the longest common subsequences were for the examples in the training and test sets.

**Comment by shwina** [comment]
@aclifton314 Could you help me with a few questions:

* Can I get a sense of how many string sequences you are working with? That is, in your N-by-N matrix, what is "N"?

* cuDF/dask-cuDF are much better suited to working with tall-and-skinny dataframes rather than square/wide DataFrames. Given that, can the problem be recast to something like the following?

Given a DataFrame with two columns `x` and `y`, each containing sequences of strings:

```
                a                b
0       aaa bb cc        bb aab cc
1  abc abd ab xyz  abd abcd ab xyz
```

Compute a third column `z` that contains the largest common subsequence of `x` and `y`:

```
                a                b           c
0       aaa bb cc        bb aab cc       bb cc
1  abc abd ab xyz  abd abcd ab xyz  abd ab xyz
```

**Comment by aclifton314** [comment]
@shwina no problem:
- Here are the shapes
train: (15047, 15047)
test: (1672, 1672)

- Yes I definitely think this can be recasted. I had the problem set up this initial way just because that's the way my mind thinks about it. It could be setup something like this:
```python
         dataset                lcs
0       aaa bb cc        lcs(seq0, seq1)
1  abc abd ab xyz        lcs(seq0, seq2)
```
Ultimately, I'm wanting to run the LCS calculation over all pairs in the training set, all pairs in the test set, and all pairs between the training and test set. I'm not concerned with the reciprocals so that can cut down on some of the calculation (i..e. lcs(a, b) and not lcs(b, a)). Does this help?

**Comment by aclifton314** [comment]
Any ideas about how to do this?

**Comment by shwina** [comment]
It's an interesting problem. I did find [this paper](https://www.researchgate.net/profile/Yang-Jiaoyun/publication/45534436_An_Efficient_Parallel_Algorithm_for_Longest_Common_Subsequence_Problem_on_GPUs/links/09e4150fd67e2142d7000000/An-Efficient-Parallel-Algorithm-for-Longest-Common-Subsequence-Problem-on-GPUs.pdf) which describes an algorithm for doing very large LCS on the GPU, however, it sounds like what you have is lots of little LCS problems. Is that correct, i.e., how large on average is each of the sequences?

**Comment by quasiben** [comment]
If this is a single column, have you tried with cupy/dask+cupy instead of dask-cudf ?

**Comment by aclifton314** [comment]
@shwina thanks for your response. Yes it appears that really I'm doing a bunch of smaller LCS calculations. On average, the sequence lengths are about 60 - 70 tokes long.

@quasiben I have not tried cupy/dask+cupy. My biggest hurdle right now is setting up the syntax correctly to make these calculations run faster.

**Comment by aclifton314** [comment]
@quasiben I was also assuming that given the GPU cluster I have dask-cudf might be a good solution, but I am not 100% committed to using it if something better exists.

**Comment by vyasr** [comment]
Since as @shwina pointed out you are doing a large number of smaller LCS calculations, it seems like you don't need a particularly efficient LCS calculation, you just need to parallelize over all the calculations. If you can modify your `DataFrame` into the form he originally described:
```
                x                y
0       aaa bb cc        bb aab cc
1  abc abd ab xyz  abd abcd ab xyz
``` 
Then you should be able to just apply something like your original `lcsubseq` using `apply_rows` ([here are the docs for applying UDFs](https://docs.rapids.ai/api/cudf/stable/guide-to-udfs.html)). For example, since your training dataset has 15047 elements, you would construct a `DataFrame` of shape (15047*15047, 2) and then apply your LCS function directly to that row-wise. Since that's a pretty large frame maybe you'd want to use dask to handle it, but that approach seems simpler. It basically amounts to replacing your `make_lcsubseq_df` with simply constructing a DataFrame that contains all the pairs, then calling your `lcsubseq` function on each row (in parallel).

**Comment by aclifton314** [comment]
Thanks everybody for providing feedback and thinking through this. I don't know if it helps, but my dataset can fit in normal CPU memory.

**Comment by aclifton314** [comment]
@vyasr Would you recommend using `apply_rows` on a `dask-cudf` dataframe for the time being?

**Comment by vyasr** [comment]
That idea actually won't actually work, sorry about that (see my edit above). The problem is that inside your `lcubseq` you need to create local arrays, which numba doesn't support. I'm still trying to come up with a solution, but I unfortunately don't have a great answer for you yet.
